# IE - Master in Business Analytics and Big Data

## Recommendation Systems

## Building a Non Personalized Recommendation Engine
### By: Alain Grullón González

Import necessary libraries

In [1]:
import pandas as pd

Load the critics dataset

In [2]:
crit_df = pd.read_csv("C:/Users/alain/OneDrive/Desktop/MBD/Term 2 (Python)/7. Recommendation Engines (10)/Lab 1/Data/critics.csv")
crit_df.head()

User   Star Wars IV - A New Hope   Star Wars VI - Return of the Jedi  \
0     John                         1.0                                 5.0   
1    Maria                         5.0                                 3.0   
2    Anton                         NaN                                 NaN   
3    Roger                         NaN                                 3.0   
4  Martina                         4.0                                 3.0   

   Forrest Gump  The Shawshank Redemption  The Silence of the Lambs  \
0           2.0                       NaN                       4.0   
1           NaN                       2.0                       4.0   
2           NaN                       5.0                       2.0   
3           NaN                       NaN                       NaN   
4           4.0                       1.0                       4.0   

   Gladiator  Toy Story  Saving Private Ryan  Pulp Fiction  ...  \
0        4.0        2.0                  2.0           NaN  ...   
1        2.0        1.0                  NaN           NaN  ...   
2        NaN        4.0                  NaN           NaN  ...   
3        1.0        2.0                  3.0           4.0  ...   
4        1.0        NaN                  4.0           NaN  ...   

   Shakespeare in Love  Total Recall  Independence Day  Blade Runner   \
0                  2.0           NaN               5.0            2.0   
1                  3.0           2.0               2.0            NaN   
2                  NaN           1.0               4.0            4.0   
3                  NaN           4.0               1.0            3.0   
4                  5.0           1.0               NaN            4.0   

   Groundhog Day  The Matrix  Schindler's List  The Sixth Sense  \
0            NaN         4.0               2.0              5.0   
1            2.0         NaN               5.0              1.0   
2            1.0         1.0               2.0              3.0   
3            5.0         NaN               5.0              1.0   
4            NaN         3.0               5.0              5.0   

   Raiders of the Lost Ark  Babe  
0                      NaN   NaN  
1                      3.0   NaN  
2                      1.0   3.0  
3                      1.0   2.0  
4                      NaN   NaN  

[5 rows x 21 columns]

**1) Top 5 Movies, by rating mean**

Calculate mean rating for each movie, ordered with the highest rating listed first, and plot the top 5.

The top 5 rated movies are:

In [3]:
crit_df.mean().sort_values(ascending=False).head(5)

The Shawshank Redemption      3.600000
 Star Wars IV - A New Hope    3.266667
Blade Runner                  3.222222
Groundhog Day                 3.166667
The Silence of the Lambs      3.062500
dtype: float64

**2) Top 5 Movies, by rating distribution**

Calculate the percentage of ratings for each movie that are 4 stars or higher. Order with the highest percentage first, and plot the top 5 movies with more high ratings distribution.

In [4]:
# The moves with higher than 4 stars
top_df = (crit_df.iloc[:,1:21]>=4).sum()
# The total ratings per movie
total_df = crit_df.iloc[:,1:].count()
# Percentage of 4 stars or higher per movie
percent_df = (top_df / total_df).sort_values(ascending=False).head(5)
print(percent_df)

The Shawshank Redemption      0.700000
 Star Wars IV - A New Hope    0.533333
Gladiator                     0.500000
Blade Runner                  0.444444
The Silence of the Lambs      0.437500
dtype: float64


**3) Top 5 Movies, by quantity of ratings**

Count the number of ratings for each movie, order the one with most number of ratings first, submit the top 5.

In [5]:
# The total ratings per movie
total_df = crit_df.iloc[:,1:].count()
# Top 5 most rated
top5 = total_df.sort_values(ascending=False).head(5)
print(top5)

Toy Story                             17
The Silence of the Lambs              16
 Star Wars IV - A New Hope            15
 Star Wars VI - Return of the Jedi    14
Independence Day                      13
dtype: int64


**4) People who watched Star Wars IV also watched ...**

Calculate movies that most often occur with other movie. For each movie, calculate the percentage of the other movie raters who also rated that movie. Order with the highest percentage first, and submit the top 5. Select "Star.Wars.IV...A.New.Hope" movie by defaut. Use the (x+y)/x method.

In [6]:
# Watched Star Wars IV
watched_sw4 = crit_df[" Star Wars IV - A New Hope"].count()
# Watched other
watched_other = crit_df[" Star Wars IV - A New Hope"].notnull()
# New DataFrame with only SW4 Watchers
crit_df_SW4 = crit_df[watched_other]
# New DataFrame to represent also watched...
others_df = crit_df_SW4.drop(["User"," Star Wars IV - A New Hope"], 1)
also_watched_df = others_df.count().sort_values(ascending=False)
# Percentage of people who watched SWIV and also watched...
percentage = (also_watched_df/watched_sw4).head(5)
percentage

Toy Story                             0.933333
 Star Wars VI - Return of the Jedi    0.866667
The Silence of the Lambs              0.800000
Independence Day                      0.733333
Total Recall                          0.666667
dtype: float64

**5) People who liked Babe also liked ...**

Calculate the movies better rated of people who liked a movie. Select the people who liked the movie "Babe" (4 or 5 stars) and provide the top 5 movies they liked most.

In [8]:
# Babe Lovers
bl_df = crit_df.loc[crit_df["Babe"]>=4]
# Babe Lovers + Other movies
bl_plus_df = bl_df.drop(["User","Babe"],1)
# Best 5 from the Babe Lovers
top5_babe_plus = bl_plus_df.mean().sort_values(ascending=False).head(5)
top5_babe_plus

Groundhog Day               5.000000
Pulp Fiction                5.000000
The Shawshank Redemption    4.500000
Toy Story                   4.333333
Blade Runner                4.000000
dtype: float64

**6) MovieLens**

Explore a real data set and provide non-personalized ratings. You can use the MovieLens database. You can find MovieLens' dataset here: http://files.grouplens.org/datasets/movielens/ml-10m-README.html

Lets first read the datasets in

In [9]:
movie_df = pd.read_csv("C:/Users/alain/OneDrive/Desktop/MBD/Term 2 (Python)/7. Recommendation Engines (10)/Lab 1/Data/movies.csv")
ratings_df = pd.read_csv("C:/Users/alain/OneDrive/Desktop/MBD/Term 2 (Python)/7. Recommendation Engines (10)/Lab 1/Data/ratings.csv")

In [119]:
movie_df.count()

movieId    58098
title      58098
genres     58098
dtype: int64

58,098 movies in this dataset

In [120]:
movie_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

Lets delete the genres column since it is of no interest.

In [11]:
mov_df = movie_df.drop("genres",1)

What about ratings

In [121]:
ratings_df.count()

userId       27753444
movieId      27753444
rating       27753444
timestamp    27753444
dtype: int64

27,753,444 ratings

In [122]:
ratings_df.head()

userId  movieId  rating   timestamp
0       1      307     3.5  1256677221
1       1      481     3.5  1256677456
2       1     1091     1.5  1256677471
3       1     1257     4.5  1256677460
4       1     1449     4.5  1256677264

Lets now drop the timestamp column 

In [13]:
rat_df = ratings_df.drop("timestamp",1)

Now, lets join these two tables together to have movie and rating in the same dataset

In [95]:
merged_df = rat_df.merge(mov_df, on='movieId', how='left')

In [123]:
merged_df.count()

userId     27753444
movieId    27753444
rating     27753444
title      27753444
dtype: int64

27 Million rows, good this means the merge worked.

Again, dropping redudant columns

In [128]:
merg_df = merged_df.drop(["movieId"],1)
merg_df.tail(10)

userId  rating                                              title
27753434  283228     4.5                       Night at the Opera, A (1935)
27753435  283228     4.5                        To Have and Have Not (1944)
27753436  283228     5.0                            Another Thin Man (1939)
27753437  283228     5.0                          After the Thin Man (1936)
27753438  283228     5.0  Maltese Falcon, The (a.k.a. Dangerous Female) ...
27753439  283228     4.5                         Day at the Races, A (1937)
27753440  283228     4.5                            My Favorite Wife (1940)
27753441  283228     4.5                                    Serenity (2005)
27753442  283228     4.5                                       Brick (2005)
27753443  283228     4.5                       Bourne Ultimatum, The (2007)

**1) The best rated movies!**

In [99]:
# Grouping by titles
grouped_df = merg_df.groupby('title').mean().reset_index()
grouped_df.head()

title         userId    rating
0  "Great Performances" Cats (1998)  140307.200000  2.815556
1        #1 Cheerleader Camp (2010)  172008.444444  2.777778
2                  #Captured (2017)  179605.000000  2.500000
3                    #Horror (2015)  147384.000000  2.078947
4                 #SCREAMERS (2016)  257888.000000  2.500000

Top Rated Movies

In [143]:
# Top 5 rated movies
group_df = grouped_df.sort_values(by="rating", ascending=False)
group_df.head(5)

title    userId  rating
50501                 VHS Revolution (2017)  254422.0     5.0
51532               We Are Mountains (1969)   74049.0     5.0
13333  Dragon Lee Vs. The 5 Brothers (1978)  229337.0     5.0
26316                 Living on Love (1937)  225652.0     5.0
8644                 Change of Plans (2011)  132948.0     5.0

In [146]:
group_df.count()

title     53817
userId    53817
rating    53817
dtype: int64

53,817 rated movies

In [149]:
fivestar_df = group_df.loc[group_df["rating"]==5.0]
fivestar_df.count()

title     536
userId    536
rating    536
dtype: int64

There are 536 movies with an average rating of 5 stars.

I would build the non personalized recommendation engine on these 536 movies!